### PR TITLE
Better BPINN Log Posterior formulation and Experiments

### PR DESCRIPTION
**Pull request by AstitvaAggarwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context
Uses extra loss term

### CODE REVIEW

**Comment by AstitvaAggarwal** [comment]
@ChrisRackauckas, @Vaibhavdixit02 this has code for both PDE and ODE case over Master Branch.

**Comment by ChrisRackauckas** [comment]
What do you mean?

There's some conflicts here.

**Comment by AstitvaAggarwal** [comment]
the extra loss funciton in case of the ODE solver and PDE solver, im still working on the PDE solver case. Ill fix the conflicts as im currently focusing on this pr

File: src/PDE_BPINN.jl [hunk 1]
```diff
@@ -57,23 +58,89 @@ mutable struct PDELogTargetDensity{
...
+    # masking operation
+    eqs_new = substitute.(eqs, Ref(Dict_differentials))
+
+    to_subs, tobe_subs = get_symbols(dataset, depvars, eqs)
+    # for each row in dataset create u values for substituting in equation, n_equations=n_rows
+    eq_subs = [Dict(tobe_subs[depvar] => to_subs[depvar][i] for depvar in depvars)
+               for i in 1:size(dataset[1][:, 1])[1]]
+
+    # for each point(eq_sub dictionary) substiute in all equations(eqs_new - masked equations)
+    b = []
```
Comment (Vaibhavdixit02) [inline src/PDE_BPINN.jl:84]:
```suggestion
    b = [[substitute(eq, eq_sub) for eq in eqs_new] for eq_sub in eq_subs]
```

File: src/PDE_BPINN.jl [hunk 2]
```diff
@@ -57,23 +58,89 @@ mutable struct PDELogTargetDensity{
...
+    for eq_sub in eq_subs
+        push!(b, [substitute(eq, eq_sub) for eq in eqs_new])
+    end
+    # now we have vector of equation vectors
+
+    # reverse dict for re-substituting values of Differential(t)(u(t)) etc
+    rev_Dict_differentials = Dict(value => key for (key, value) in Dict_differentials)
+
+    # for each vector in vector of equation vectorbroadcast resubstituing OG mask values
+    c = []
```
Comment (Vaibhavdixit02) [inline src/PDE_BPINN.jl:94]: Same as above

File: src/training_strategies.jl [hunk 3]
```diff
@@ -14,20 +14,48 @@ struct GridTraining{T} <: AbstractTrainingStrategy
...
+    depvars = pinnrep.depvars
+    dict_depvars = pinnrep.dict_depvars
+    dict_indvars = pinnrep.dict_indvars
+
+    symbols_input = [(i, dict_depvar_input[i]) for i in depvars]
+    # [(:u, [:t])]
+    eq_args = NeuralPDE.get_argument(eqs, dict_indvars, dict_depvars)
+    # [[:t]]
+
+    points = []
```
Comment (Vaibhavdixit02) [inline src/training_strategies.jl:28]: Type it or create with comprehension

File: src/training_strategies.jl [hunk 4]
```diff
@@ -14,20 +14,48 @@ struct GridTraining{T} <: AbstractTrainingStrategy
...
+    for eq_arg in eq_args
+        a = []
```
Comment (Vaibhavdixit02) [inline src/training_strategies.jl:30]: same

**Comment by Vaibhavdixit02** [comment]
I think broadly this looks good already. Do you have any results for comparison of the new likelihood?

**Comment by ChrisRackauckas** [comment]
Agreed I look at this and it seems ready-ish to merge. What's the status here?

**Comment by AstitvaAggarwal** [comment]
Im still running experiments, I've managed to get one test set pass for the new implementation but that's a simple example, the Lorentz equations example fails as I still need to tune the training process better. Overall im working on the new test problems

**Comment by Vaibhavdixit02** [comment]
Okay sounds good. What do you mean by failing for the Lorenz one? Also Lorenz is probably way too hard to be the test problem, use the LV instead

**Comment by AstitvaAggarwal** [comment]
I mean the estimated parameters are not within tolerated error bounds, ive not been able to get HMC or NUTS reduce the Hamiltonian enough in all the runs till now. Yeah, will try the same for KS and LV now

**Review by Vaibhavdixit02 — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
You may want to rebase to master so it should go faster.

**Comment by AstitvaAggarwal** [comment]
@ChrisRackauckas i apologise for the delayed PR, ive been testing the implementation thoroughly. ive shared results uptil now with @Vaibhavdixit02. The results are positive but not significant enough, this is for the LV case as the PDE residual loss is misleading (allows zero gradients trivial solution for inverse problem setting). this PR involves the new loss implementation, some tests, and a correction to the existing PDE BPINN solver(was hidden in the details of the Domain point generation function present in the normal PINN solver). I will add more tests in a seperate PR.

**Comment by AstitvaAggarwal** [comment]
image
green and pink curve for new implementation(sol5_2), blue and brown for old implementation(sol3)
image

**Pull request closed without merging**